src/build_system.ml

```diff
@@ -1173,8 +1174,22 @@ let eval_request t ~request ~process_target =
        let dyn_deps = Build_exec.exec_nop t request () in
        process_targets (Pset.diff dyn_deps static_deps))
 
+let universe_file = Path.relative Path.build_dir ".universe-state"
+
+let update_universe t =
+  let fname = Path.to_string universe_file in
+  let n =
+    if Sys.file_exists fname then
+      Sexp.Of_sexp.int (Sexp.load ~mode:Single ~fname) + 1
+    else
+      0
+  in
+  make_local_dirs t (Pset.singleton Path.build_dir);
+  Io.write_file fname (Sexp.to_string (Sexp.To_sexp.int n))
+
 let do_build t ~request =
   entry_point t ~f:(fun () ->
+    update_universe t;
```

Comment:
It's a bit of a shame that we can't do this update a bit more lazily. I.e. only keep this universe file updated once we actually detect dependencies on it. I don't see an easy way to do it though, so it's not worth the hassle.

Reply:
We could add a special rule for this file, but we already write a few files on every run, so it doesn't seem worth it indeed